### PR TITLE
API Tests

### PR DESCRIPTION
**Pull request by AustinShalit**

### CODE REVIEW

**Comment by AustinShalit** [comment]
@JLLeitschuh Is there something I can do better with the build files?

File: api/src/main/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtils.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
   private static final Pattern newMetadataPattern = Pattern.compile("/\\.");
 
   private NetworkTableUtils() {
+    throw new UnsupportedOperationException("This is a utility class!");
```
Comment (nightpool) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtils.java]: the private constructor is sufficient here.

File: api/src/test/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                 "false, /~METADATA",
+                 "false, /METADATA~",
+                 "true, /~METADATA~",
+                 "true, /.metadata",
+                 "true, /~METADATA~/someOtherValue",
+                 "true, /.metadata/someOtherValue",
+                 "false, /my.key.with.dots",
+                 "false, /my~key~with~tildes~",
+                 "false, /~my~keywithtildes",
+                 "false, /~metadata~with~tildes~"})
```
Comment (nightpool) [inline api/src/test/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtilsTest.java]: this line is an edge case and I don't want to decide on it either way.

File: build.gradle.kts [hunk 3]
```diff
@@ -45,7 +46,13 @@ subprojects {
                 create(group = "org.junit.jupiter", name = name, version = version)
         "testCompile"(junitJupiter(name = "junit-jupiter-api"))
         "testCompile"(junitJupiter(name = "junit-jupiter-engine"))
+        "testCompile"(junitJupiter(name = "junit-jupiter-params"))
         "testRuntime"(create(group = "org.junit.platform", name = "junit-platform-launcher", version = "1.0.0-M5"))
+        fun testFx(name: String, version: String = "4.0.+") =
+                create(group = "org.testfx", name = name, version = version)
+        "testCompile"(testFx(name = "testfx-core"))
+        "testCompile"(testFx(name = "testfx-junit5"))
```
Comment (nightpool) [inline build.gradle.kts]: seems unused?

File: api/src/main/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtils.java (hunk 1 above)
Comment (AustinShalit) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtils.java]: This prevents instantiation from reflection

File: api/src/test/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtilsTest.java (hunk 2 above)
Comment (nightpool) [inline api/src/test/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtilsTest.java]: this line is an edge case and I don't want to decide on it either way.

File: build.gradle.kts (hunk 3 above)
Comment (AustinShalit) [inline build.gradle.kts]: Going to be used once this PR is not WIP.

File: api/src/main/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtils.java (hunk 1 above)
Comment (nightpool) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtils.java]: why do we care if someone instantiates with reflection?

Comment (SamCarlberg) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/util/NetworkTableUtils.java]: It prevents instantiation from within the class. Useful from a maintainability standpoint

**Comment by AustinShalit** [comment]
On hold until #144 is solved

File: api/src/test/java/edu/wpi/first/shuffleboard/api/util/EqualityUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+public class EqualityUtilsTest extends UtilityClassTest<EqualityUtils> {
+
+  public EqualityUtilsTest() {
+    super(EqualityUtils.class);
+  }
+
+  @Test
+  public void isEqualInvalidPrimativeTest() {
+    assertThrows(UnsupportedOperationException.class, () -> EqualityUtils.isEqual(new long[0], new long[0]));
```
Comment (AustinShalit) [inline api/src/test/java/edu/wpi/first/shuffleboard/api/util/EqualityUtilsTest.java]: @SamCarlberg Are long arrays something this method should support?

File: api/src/test/java/edu/wpi/first/shuffleboard/api/util/UtilityClassTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SuppressWarnings("PMD.AbstractClassWithoutAbstractMethod")
+public abstract class UtilityClassTest<T> {
+
+  private final Class<T> clazz;
+
+  public UtilityClassTest(Class<T> clazz) {
+    this.clazz = clazz;
+  }
```
Comment (JLLeitschuh) [inline api/src/test/java/edu/wpi/first/shuffleboard/api/util/UtilityClassTest.java]: You could simplify this constructor by removing the constructor argument and doing this instead:
```java
public UtilityClassTest() {
    this.clazz = (new TypeToken<T>(getClass()) {}).getRawType();
}
```

File: api/src/test/java/edu/wpi/first/shuffleboard/api/util/UtilityClassTest.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  @Test
+  public void testConstructorPrivate() throws NoSuchMethodException {
+    Constructor<T> constructor = clazz.getDeclaredConstructor();
+
+    assertFalse(constructor.isAccessible());
```
Comment (JLLeitschuh) [inline api/src/test/java/edu/wpi/first/shuffleboard/api/util/UtilityClassTest.java]: A message would be useful here.

File: api/src/test/java/edu/wpi/first/shuffleboard/api/util/UtilityClassTest.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  @Test
+  public void testConstructorReflection() {
+    assertThrows(InvocationTargetException.class, () -> {
+      Constructor<T> constructor = clazz.getDeclaredConstructor();
+      constructor.setAccessible(true);
+      constructor.newInstance();
+    });
```
Comment (JLLeitschuh) [inline api/src/test/java/edu/wpi/first/shuffleboard/api/util/UtilityClassTest.java]: A message would be useful here.

File: app/app.gradle.kts [hunk 8]
```diff
@@ -60,11 +55,11 @@ if (project.hasProperty("jenkinsBuild") || project.hasProperty("headless")) {
...
-                "-Djava.awt.headless=true",
-                "-Dtestfx.robot=glass",
-                "-Dtestfx.headless=true",
-                "-Dprism.order=sw",
-                "-Dprism.text=t2k"
+                    "-Djava.awt.headless=true",
+                    "-Dtestfx.robot=glass",
+                    "-Dtestfx.headless=true",
+                    "-Dprism.order=sw",
+                    "-Dprism.text=t2k"
```
Comment (JLLeitschuh) [inline app/app.gradle.kts]: Indentation changes?

Comment (AustinShalit) [inline app/app.gradle.kts]: I think my IDE and @SamCarlberg's IDE are fighting over the indentation of this file.  Will #146 fix this?

Comment (JLLeitschuh) [inline app/app.gradle.kts]: No, the indentation rule isn't correctly implemented yet.
https://github.com/shyiko/ktlint/issues/76

**Comment by AustinShalit** [comment]
@nightpool I'd like to merge this PR and then open another one with more progress.  That way we have coverage for PRs.

File: api/api.gradle.kts [hunk 9]
```diff
@@ -14,3 +14,32 @@ dependencies {
...
+            jvmArgs = listOf(
+                    "-Djava.awt.headless=true",
+                    "-Dtestfx.robot=glass",
+                    "-Dtestfx.headless=true",
+                    "-Dprism.order=sw",
+                    "-Dprism.text=t2k"
+            )
+        }
+    }
+}
```
Comment (JLLeitschuh) [inline api/api.gradle.kts]: Why not move this logic to the root `build.gradle.kts` file?

File: plugins/base/src/test/java/edu/wpi/first/shuffleboard/plugin/base/widget/BooleanBoxTest.java [hunk 10]
```diff
@@ -48,6 +49,7 @@ public void testColorWhenFalse() {
   }
 
   @Test
+  @Disabled
```
Comment (JLLeitschuh) [inline plugins/base/src/test/java/edu/wpi/first/shuffleboard/plugin/base/widget/BooleanBoxTest.java]: A reason would have been helpful here.
